presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java

```diff
@@ -0,0 +1,644 @@
+        }
+
+        LogicalPropertiesImpl build()
+        {
+            keyProperty.addKeys(sourceProperties.keyProperty);
+            maxCardProperty.update(sourceProperties.maxCardProperty);
+            equivalenceClassProperty.update(sourceProperties.equivalenceClassProperty);
+            if (equivalenceClassProperty.update(predicate)) {
+                normalizeKeyPropertyAndSetMaxCard(keyProperty, maxCardProperty, equivalenceClassProperty);
+            }
+            return new LogicalPropertiesImpl(equivalenceClassProperty, maxCardProperty, keyProperty);
+        }
+    }
+
+    /**
+     * This logical properties builder should be used by PlanNode's that project their
+     * source properties. For example, a ProjectNode and AggregationNode project their
+     * source properties. The former might also reassign property variable references.
+     */
+    public static class ProjectBuilder
+    {
+        private final LogicalPropertiesImpl sourceProperties;
+        private final Assignments assignments;
+        private final MaxCardProperty maxCardProperty = new MaxCardProperty();
+        private KeyProperty keyProperty = new KeyProperty();
+        private EquivalenceClassProperty equivalenceClassProperty;
+
+        ProjectBuilder(LogicalPropertiesImpl sourceProperties, Assignments assignments, FunctionResolution functionResolution)
+        {
+            requireNonNull(sourceProperties, "sourceProperties is null");
+            requireNonNull(assignments, "assignments is null");
+            this.sourceProperties = sourceProperties;
+            this.assignments = assignments;
+            this.equivalenceClassProperty = new EquivalenceClassProperty(functionResolution);
+        }
+
+        LogicalPropertiesImpl build()
+        {
+            keyProperty.addKeys(sourceProperties.keyProperty);
+            maxCardProperty.update(sourceProperties.maxCardProperty);
+            equivalenceClassProperty.update(sourceProperties.equivalenceClassProperty);
+
+            //project both equivalence classes and key property
+            Map<VariableReferenceExpression, VariableReferenceExpression> inverseVariableAssignments = inverseVariableAssignments(assignments);
+            keyProperty = keyProperty.project(new InverseVariableMappingsWithEquivalence(equivalenceClassProperty, inverseVariableAssignments));
+            equivalenceClassProperty = equivalenceClassProperty.project(inverseVariableAssignments);
+            normalizeKeyPropertyAndSetMaxCard(keyProperty, maxCardProperty, equivalenceClassProperty);
+            return new LogicalPropertiesImpl(equivalenceClassProperty, maxCardProperty, keyProperty);
+        }
+    }
+
+    /**
+     * This logical properties builder should be used by PlanNode's that propagate their
+     * source properties and add a limit. For example, TopNNode and LimitNode.
+     */
+    public static class PropagateAndLimitBuilder
```

Comment:
None of these Builder classes are actual builders. They don't need to exist. Just have static creator methods like `public static LogicalPropertiesImpl propagateAndLimitProperties(...)`, `public static tableScanProperties(...)`, etc.